[RHDEVDOCS-2879] Document ability to customize available cluster roles for Projects (#40054)
openshift/openshift-docs: modules/odc-customizing-available-cluster-roles-using-developer-perspective.adoc

// Module included in the following assemblies:
//
// applications/projects/working-with-projects.adoc

[id="odc-customizing-available-cluster-roles-using-developer-perspective_{context}"]
= Customizing the available cluster roles using the Developer perspective

> Customize the cluster role options of an existing project mentioned in the *Project* > *Project access* tab > *Role* drop down, by editing the YAML code in the `console.openshift.io/display-name: v1`.

Reviewer: This abstract seems more like a procedure step to me. 

How about giving a one sentence overview of why they'd customize the cluster role options? Something like:

`To have more control over your projects and to make changes to your project based off your needs, customize the cluster role options of an existing project.` 

I'd then scratch this last sentence from the abstract and make it step one of the procedure. 

1. Select  *Project* > *Project access* tab > *Role* drop down and edit the  YAML code in the `console.openshift.io/display-name: v1`.
2. Search for "*Console*" in the *Search* view > *Resources* drop down.
3. So on and so forth.

Author: Sounds fair. Let me rephrase a bit more based on your example.

> Each user of a project is assigned to a cluster role based on their access control. These cluster roles are mentioned in the *Project* view > *Project access* tab > *Role* drop down. By default these roles are *Admin*, *Edit*, and *View*. To add or edit the cluster roles for a project, you can customize the YAML code of the cluster.

Reviewer: See grammatical changes: 

The users of a project are assigned to a cluster role based on their access control. These cluster roles are mentioned in the *Project* view > *Project access* tab > *Role* drop down. By default, these roles are *Admin*, *Edit*, and *View*. To add or edit the cluster roles for a project, you can customize the YAML code of the cluster.

> The users of a project are assigned to a cluster role based on their access control. These cluster roles are mentioned in the *Project* view > *Project access* tab > *Role* drop down. By default, these roles are *Admin*, *Edit*, and *View*. To add or edit the cluster roles for a project, you can customize the YAML code of the cluster.

Author: Done

.Project roles
image::odc_cluster_role.png[]

> 

Reviewer 2: Suggest dropping this screenshot from maintenance pov. The instructions are clear enough. Also, the previous module has a similar screenshot, though that might not be relevant when we consider each module to be self sufficient. But like I said the instructions are clear in this case.

Author: Agreed with your point on this. Let me remove the screenshot from here.

.Procedure
To customize the different cluster roles of a project:

> . Search for "Console" in the *Search* view > *Resources* drop down.

Reviewer 2: When writing for GUI, location then action.
Suggest: In the *Search* view, use the *Resources* drop-down list to search for `Console`.
User defined items use backticks. We never use quotation marks in our docs. Please see the Asciidoc markup section in the OCP guidelines on GH: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference

> . Select the *Console `operator.openshift.io/v1`*.

Reviewer 2: Suggest: From the available options, select...

Author: Done

// Module included in the following assemblies:
//
// applications/projects/working-with-projects.adoc

[id="odc-customizing-available-cluster-roles-using-developer-perspective_{context}"]
= Customizing the available cluster roles using the Developer perspective

> You can customize cluster role's name of existing projects by editing the YAML code in the `console.openshift.io/display-name: v1`.

Reviewer: Some missing articles (see my rephrase). Also, what's the point of customizing the cluster role's name? See my rephrase to give a very quick, simple explanation of why users would care to customize the cluster role's name:

`Customize the cluster role's name of existing projects by editing the YAML code in the `console.openshift.io/display-name: v1`. When you customize the cluster role's name, you can <state benefit>.`

Author: The first part of the suggestion aligns with what is expected on the page. The other part is not required. So I'll rephrase accordingly.

.Procedure

> Customize the cluster role options of all projects that are shown in the *Project Access* tab:

Reviewer: Above, you say to customize the cluster role's name. Here, you say to customize the cluster role's options. Which is it? The title says "roles" so perhaps here you can say "role" instead of "option" for consistency? Unless they are two different things?

Also, you sometimes use possessive: cluster role's name. 
Other times, you don't use a possessive: cluster role options. 

Pick one to be consistent. I would go with whichever case is more common in openshift docs. If it helps, grammatically, it would be possessive since both the name and options belong to the cluster role.

Author: Good point, I'll keep this suggestion while rephrasing the content.

Reviewer 3: @jc-berger Thank you, Jake, for looking into this.
It was one of my concerns too: the introductory paragraph seems a bit confusing to me since it describes some fine-tuning option (customizing roles names) while it's not this module's main topic. I hope you could help us solve this, thanks.

> . In the *Search* view, select the *Resources* drop down and search "*Console*".

Reviewer: Your previous sentence says that we want to be in the *Project Access* tab. In your first step, you tell me to go to the *Search* view -> *Resources* -> *Console* So am I supposed to be in the Console or Project Access tab? Or is the Console within the Project Access tab?

Honestly, I'm not even sure if you need the first sentence before the procedure, especially since it can cause confusion as to where to direct the user. I think if you make the use case clearer in the abstract, you can simply start with step 1 under the procedure. So: 

**Title**
<abstract>
**Procedure**
1....

Author: Sounds good. Will try to simplify it a bit more.

. Select the *Console `operator.openshift.io/v1`*.
+
.Searching Console resource

> image::odc_cluster_console.png[]

Reviewer: Though I find the screenshot helpful, I know they can be quite difficult for accessibility, a growing concern in RH documentation. They also are hard to maintain up to date. 

Personally, I find it best to avoid them. However, if your team for this specific document believes the screenshot would help all users, keep it.

Reviewer: If you do keep the screenshot, end step two with a colon, not a period, since it introduces the screenshot:

2. Select the Console operator:

Author: In ODC we keep the sentence closed with a period. For reference check this page: https://docs.openshift.com/container-platform/4.9/applications/odc-viewing-application-composition-using-topology-view.html#odc-interacting-with-applications-and-components_viewing-application-composition-using-topology-view

> . Select *cluster* under the *Name* list.

Reviewer: Verify that cluster isn't capitalized. In the screenshot, I see all the UI elements are capitalized, so "cluster" here very well may need to be capitalized as well.

Author: I'll verify it again.

Author: The "cluster" is not capitalized in the UI. Check this:

![image](https://user-images.githubusercontent.com/3268242/152976356-6c56e2dc-f75d-4398-b559-695562770c53.png)

Reviewer 2: Principle of location before action: Suggest: Under the *Name* list, select...
The user needs to know where the field is before he/she takes action on it.

> . Navigate to *YAML* tab to view and edit the YAML code.

Reviewer: Navigate to _the_ *YAML* tab..

Author: Done.

> . In the YAML code under `spec` add or edit the list of `availableClusterRoles`.

Reviewer: Since this step proceeds the yaml snippet, end it with a colon.

> . In the YAML code under `spec` add or edit the list of `availableClusterRoles`.

Author: Sure.

+
[source,yaml]
----
spec:
  customization:
    projectAccess:
      availableClusterRoles:
      - admin
      - edit
      - view
----
+
For example, you can remove `view` role from the list to show `admin` and `edit` roles only.

> . Click *Save*.

Reviewer: I think this step could be combined with the previous step. Something like this: 

`In the YAML code under `spec`, add or edit the list of `availableClusterRoles` and save your changes:
YAML snippet.

Author: Sounds like a good suggestion. I'll try this one too.

> Verify the changes in the *Role* drop down under the *Project Access* tab of *Project* view.

Reviewer: Make this a verification section:

.Verification 

- To verify your changes, go to the *Project* view > *Project Access* Tab > *Role* drop-down and confirm your changes have been made. 

Also, with these instructions being fairly simple, I don't think the screenshot is necessary. Granted, I'm not the biggest fan of screenshots, so I might be biased :)

Author: Thanks for the suggestion. I'll check whether do we keep `.Verficiation` step in ODC or not as I never encountered one so far. For the screenshot I might wanna keep it as it helps users to understand things better.

> .Project roles

Reviewer: I don't see the project roles within this procedure in the preview link.

Author: Those are there in the screenshot. Let us check again.

// Module included in the following assemblies:
//
// applications/projects/working-with-projects.adoc

[id="odc-customizing-available-cluster-roles-using-developer-perspective_{context}"]
= Customizing the available cluster roles using the Developer perspective

> You can customize and configure an existing project access roles name by editing the YAML code of `console.openshift.io/display-name` of your cluster.

Reviewer 4: Small details:
* Does it make sense to note here again that you need to changing the `ClusterRole`? But we are in the right section, so it is maybe obvious.
* You should mention that this is an `annotation`, otherwise the admin could also think you mean a label.

```suggestion
You can customize and configure an existing project access roles name by editing the YAML code, and add an `console.openshift.io/display-name: Display name` annotation of your cluster.
```

Author: Got it. This suggestion seems fine I'll rephrase the sentence a bit to incorporate the changes.

> You can customize and configure an existing project access roles name by editing the YAML code, and add an `console.openshift.io/display-name: Display name` annotation to the cluster.

Author: I think the "**access role's name**" is fine here. It is actually the list of roles that can be assigned to a user for a particular cluster.

> You can customize and configure an existing project access roles name by editing the YAML code, and add an `console.openshift.io/display-name: Display name` annotation to the cluster.

Author: @missmesss actually the YAML code resides in the `console.openshift.io/display-name: Display name` file. Where we can add/edit the existing list of user roles of a cluster. And the annotation part is to edit the list in the YAML code.

> You can customize and configure an existing project access roles name by editing the YAML code, and add an `console.openshift.io/display-name: Display name` annotation to the cluster.

Reviewer 3: @rishumehra I'm afraid I still don't quite understand why this "Display name" feature is mentioned in the introduction. If I understand @jerolimov 's comment correctly, it's an extra fine-tuning option. Does it help the users to read about some unnecessary option as the first thing in the module?

> You can customize and configure an existing project access roles name by editing the YAML code, and add an `console.openshift.io/display-name: Display name` annotation to the cluster.

Author: @missmesss Got it, so I'll put in something more appropriate like "v1" or something else to avoid such confusion.

> You can customize and configure an existing project access roles name by editing the YAML code, and add an `console.openshift.io/display-name: Display name` annotation to the cluster.

Reviewer 3: @rishumehra But do we need such details in the introduction at all? In some other comment you stated we needed an action item in the introduction to the procedure, where does this guideline come from, please?

> You can customize and configure an existing project access roles name by editing the YAML code, and add an `console.openshift.io/display-name: Display name` annotation to the cluster.

Reviewer 3: To sum up all my confusions: the current introduction includes a tip about displaying access roles names, it doesn't describe what the procedure is about.

Also, customize and configure are basically the same here, are they not?

```suggestion
You can customize cluster roles of existing projects by editing the YAML code.
```

> 

Reviewer 4: IMHO: I would recommend to move this to the end of the customization doc here. So giving the reader first an introduction how to configure the available options, and then with this hint you can fine-tune the displayed options in the dropdown. Wdyt?

Author: Thanks for the suggestion but as per the docs consistency, we define or explain the "what" part with the action item then the proper how-to procedure. This is a kind of template to create content on the docs site.

.Procedure

> To customize and configure the cluster roles of your project:

Reviewer 4: Did you think it's clear here that all projects are affected if the user changes the cluster configuration?

```suggestion
To customize and configure the cluster roles of all projects:
```

Author: In my understanding, it was only for the current project. And thanks for clarifying that it is for all the projects. I'll update the sentence.

> To customize and configure the cluster roles of all the projects:

Reviewer 4: I would recommend finetuning this here a little bit. The admin will customize the options in the tab here, not the cluster roles themselves. What about this change?

```suggestion
To customize which cluster roles options are shown for all projects in the *Project Access* tab:
```

Author: Great suggestion, I am gonna make this change.

> To customize the cluster role options of all projects that are shown in the *Project Access* tab:

Author: If I am not wrong, most of the Openshift documentation follows the "To" convention while starting the procedure section.

> . In the *Developer* perspective, navigate to the *Search* view.

Reviewer 4: Cluster admins work maybe more time in the admin perspective, we have a search there as well which works similar. Maybe we can just drop the perspective recommendation?

```suggestion
. Navigate to the *Search* view.
```

Author: Sounds fair, I can rephrase without using "**Developer** perspective".

> . In the *Search* view, select the *Resources* drop down and search *Console*.

Author: Got it, that's a good point, thanks!

> . Select the *Console `operator.openshift.io/v1`* to add as a new row in resource list.

Author: Thanks for pointing this.

> . Select the *Console `operator.openshift.io/v1`* to add as a new row in resource list.

Reviewer 3: @rishumehra Now that you've provided the preview of a rendered doc with the screenshots, I'm afraid I got confused by this line a bit.
It looks like the step suggests adding a new row that will be "Console `operator.openshift.io/v1". Am I misreading it? Does my suggestion make sense?

```suggestion
. Select *Console `operator.openshift.io/v1`* in the Resources list (to add a new role? Is this tail needed?)
```

Author: @missmesss Let me demo it to you for this one, then we can discuss this further. Because now I am also not sure about it.

Reviewer 3: @rishumehra Do you mean some video demo or presentation? We could save you some time as anyway I don't need to know how it works in detail. It'll be great if you check that this line is accurate, that'll be enough, thank you.

Author: @missmesss 

Here is a procedure overview for your reference:

- Actually in *Resource* tab we search for "Console" and select the "Console `operator.openshift.io/v1" resource. Like this: 
  ![image](https://user-images.githubusercontent.com/3268242/151935077-28977e81-9c67-4b44-b850-4173bbf51f0e.png)

- Then add ""Console `operator.openshift.io/v1"" item to a resource list, like this:
 ![image](https://user-images.githubusercontent.com/3268242/151935715-658198cd-8cd7-4622-bfc4-7879de58587d.png)

- Then inside it we edit the YAML code to edit the list of user access roles.
  ![image](https://user-images.githubusercontent.com/3268242/151935913-713ddccc-1704-4e64-baba-65c016a49361.png)

Author: To simplify I'll just keep: 

. Select the **Console `operator.openshift.io/v1`**.

+

> .Searching Console resource

Author: Yes, this is a title for the image. I've updated the preview link in the description section please check.

image::odc_cluster_console.png[]

> . Select *cluster* under *Name* list.

Author: Done

. Navigate to *YAML* tab to view and edit the YAML code.

> . In the YAML code under `spec` add or edit the list of `availableClusterRoles` to configure the project roles.

Reviewer 3: ```suggestion
. In the YAML code under `spec` add or edit the list of `availableClusterRoles`.
```
If I understand it correctly, this whole procedure is about configuring/customizing project roles, so repeating it inside the procedure isn't necessary maybe?

Author: That's a good point too. I'll try to rephrase.
